### PR TITLE
FAT2-156_add_search_by_postcode

### PR DESCRIPTION
**Pull request by robyoung91**

### CODE REVIEW

File: src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheGetLocationByFullPostcodeRequest.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            var actual = new GetLocationByFullPostcodeRequest(fullPostcode)
+            {
+                BaseUrl = baseUrl
+            };
+
+            actual.GetUrl.Should().Be($"{baseUrl}api/postcodes?postcode={fullPostcode}");
+        }
+
+        [Test, AutoData]
+        public void Then_The_Request_Is_Correctly_Encoded_For_Special_Characters(string baseUrl, string fullPostcode)
```
Comment (PeterDavidWilliams) [inline src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheGetLocationByFullPostcodeRequest.cs:23]: This test doesn't guarantee that there's a special character in the url. Can you modify `fullPostcode` to include a special character requiring encoding?

File: src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheLocationQueryRequest.cs [hunk 2]
```diff
@@ -15,7 +16,19 @@ public void Then_The_Request_Is_Correctly_Built(string baseUrl, string query)
                 BaseUrl = baseUrl
             };
 
-            actual.GetUrl.Should().Be($"{baseUrl}api/locations/search?query={query}");   
+            actual.GetUrl.Should().Be($"{baseUrl}api/search?query={query}");   
+        }
+
+        [Test, AutoData]
+        public void Then_The_Request_Is_Correctly_Encoded_For_Special_Characters(string baseUrl, string query)
```
Comment (PeterDavidWilliams) [inline src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheLocationQueryRequest.cs:23]: This test also doesn't guarantee a special character needing encoding, see above

**Review by PeterDavidWilliams — APPROVED** [review]

File: src/SFA.DAS.FindApprenticeshipTraining/Application/TrainingCourses/LocationHelper.cs [hunk 3]
```diff
@@ -18,13 +19,23 @@ public LocationHelper (ILocationApiClient<LocationApiConfiguration> locationApiC
...
+            }
+
+            if (location.Split(",").Length == 2)
             {
                 var locationInformation = location.Split(",");
                 var locationName = locationInformation.First().Trim();
                 var authorityName = locationInformation.Last().Trim();
                 return await _locationApiClient.Get<GetLocationsListItem>(new GetLocationByLocationAndAuthorityName(locationName, authorityName));
             }   
+            if (Regex.IsMatch(location, @"^[A-Za-z]{1,2}\d[A-Za-z\d]?\s*\d[A-Za-z]{2}$"))
```
Comment (dashton82) [inline src/SFA.DAS.FindApprenticeshipTraining/Application/TrainingCourses/LocationHelper.cs:35]: either pull that regex into a variable that describes the check, or pull the whole check into a method describing what its doing

File: src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheLocationQueryRequest.cs [hunk 4]
```diff
@@ -15,7 +16,19 @@ public void Then_The_Request_Is_Correctly_Built(string baseUrl, string query)
                 BaseUrl = baseUrl
             };
 
-            actual.GetUrl.Should().Be($"{baseUrl}api/locations/search?query={query}");   
+            actual.GetUrl.Should().Be($"{baseUrl}api/search?query={query}");   
+        }
+
+        [Test, AutoData]
```
Comment (dashton82) [inline src/SFA.DAS.FindApprenticeshipTraining.UnitTests/InnerApi/Requests/WhenBuildingTheLocationQueryRequest.cs:22]: need some inlineautodata here with some special chars

File: src/SFA.DAS.FindApprenticeshipTraining.Api.UnitTests/Models/WhenCastingGetLocationSearchResponseFromMediatorType.cs [hunk 5]
```diff
@@ -12,10 +12,13 @@ public class WhenCastingGetLocationSearchResponseFromMediatorType
...
+
             var actual = (GetLocationSearchResponse) source;
             
-            actual.Locations.ToList().First().Should().BeEquivalentTo(source.Locations.First());
-            actual.Locations.ToList().Last().Should().BeEquivalentTo(source.Locations.Last());
+            actual.Locations.ToList().First().Should().BeEquivalentTo(source.Locations.First(), options=> options.ExcludingMissingMembers());
+            actual.Locations.ToList().Last().Should().BeEquivalentTo(source.Locations.Last(), options=> options.ExcludingMissingMembers());
             actual.Locations.Count().Should().Be(source.Locations.Count());
         }
     }
```
Comment (dashton82) [inline src/SFA.DAS.FindApprenticeshipTraining.Api.UnitTests/Models/WhenCastingGetLocationSearchResponseFromMediatorType.cs:24]: Could do with some tests/assertions showing that the Name is a concatenation of the LocationName and LocalAuthorityName when the postcode is null, and also for using the postcode when that is not null or empty string

**Review by dashton82 — CHANGES_REQUESTED** [review]

**Review by dashton82 — DISMISSED** [review]

**Review by dashton82 — APPROVED** [review]

**Pull request merged**